javascript/diff_match_patch_uncompressed.js

```diff
@@ -1776,12 +1776,23 @@ diff_match_patch.prototype.patch_addContext_ = function(patch, text) {
 
   // Add the prefix.
   var prefix = text.substring(patch.start2 - padding, patch.start2);
+  
+  if(this.isLowSurrogate(prefix[0])) {
+    prefix = text.substring(patch.start2 - padding - 1 , patch.start2);
+  }
```

Comment:
just a note here: we can avoid a double string copy if we check for the surrogate half before substringing.

```js
// Avoid splitting on non-character boundaries when pulling in context.
var prefix = this.isLowSurrogate(text[patch.start2 - padding])
	? text.substring(patch.start2 - padding - 1, patch.start2)
	: text.substring(patch.start2 - padding, patch.start2);
```

Reply:
Good point, thank you, I will change this tomorrow, along with the cleanup i mentioned before. I'm off to bed for now :)

Reply:
@dmsnell This branch (and PR) was not a correct place to put my fix - I moved it over here: https://github.com/feedyou-ai/diff-match-patch/pull/4

Reply:
@dmsnell Just a quick followup: I had to also do this: https://github.com/feedyou-ai/diff-match-patch/pull/7

because `text[patch.start2 - padding]` can be out of bounds of `text` and return undefined :)